src/array/product.ts

```diff
@@ -0,0 +1,26 @@
+/**
+ * Perform a Cartesian product of arrays, combining all elements from the
+ * input arrays into all possible combinations.
+ *
+ * @see https://radashi.js.org/reference/array/product
+ * @example
+ * ```ts
+ * product([['red', 'blue'], ['big', 'small'], ['fast', 'slow']])
+ * // => [['red', 'big', 'fast'], ['red', 'big', 'slow'], ['red', 'small', 'fast'], ['red', 'small', 'slow'], ['blue', 'big', 'fast'], ['blue', 'big', 'slow'], ['blue', 'small', 'fast'], ['blue', 'small', 'slow']]
+ * ```
+ */
+export function product<T>(arrays: T[][]): T[][] {
```

Comment:
Fixing the types:

```suggestion
export function product<T extends any[]>(...arrays:  [...T]): Array<{[K in keyof T]: T[K][number]}> 
export function product<T extends any[]>(...arrays: T): T[][] {
```

Reply:
This is nothing short of brilliant, Marlon 🤯